CRM/Mailing/Event/BAO/Unsubscribe.php

```diff
@@ -149,14 +149,35 @@ public static function &unsub_from_mailing($job_id, $queue_id, $hash, $return =
     $group = $groupObject->getTableName();
     $gcObject = new CRM_Contact_BAO_GroupContact();
     $gc = $gcObject->getTableName();
+    $abObject = new CRM_Mailing_DAO_MailingAB();
+    $ab = $abObject->getTableName();
 
     //We Need the mailing Id for the hook...
     $do->query("SELECT $job.mailing_id as mailing_id
                      FROM   $job
                      WHERE $job.id = " . CRM_Utils_Type::escape($job_id, 'Integer'));
     $do->fetch();
     $mailing_id = $do->mailing_id;
+    $mailing_type = CRM_Core_DAO::getFieldValue('CRM_Mailing_DAO_Mailing', $mailing_id, 'mailing_type', 'id');
     $entity = CRM_Core_DAO::getFieldValue('CRM_Mailing_DAO_MailingGroup', $mailing_id, 'entity_table', 'mailing_id');
+
+    // If $entity is null and $mailing_Type is either winner or experiment then we are deailing with an AB test
+    $abtest_types = array('experiment', 'winner');
+    if (empty($entity) && in_array($mailing_type, $abtest_types)) {
```

Comment:
For me - `mailing_group` api [called here](https://github.com/civicrm/civicrm-core/blob/master/CRM/Mailing/BAO/Mailing.php#L1861) always ensured `entity_table` to set to either group or mailing. And if it doesn't in your case, not sure but shouldn't the fix be more likely to set this column instead of fixing the query if it is empty ?